### PR TITLE
Added DOCIS and VOICE information for Technicolor modems

### PR DESCRIPTION
**Pull request by sven-ruess**
- Reworked library_test to use local Python module
- Unified outputs for module

### CODE REVIEW

**Comment by chemelli74** [comment]
Hi @sven-ruess, thx for the contribution.
Please always install and run "`pre-commit`" before submitting a PR ;-)

If you use the built-in devcontainer, then the environment setup is done automatically

**Comment by sven-ruess** [comment]
Hi @chemelli74, what built-in devcontainer? I tried to setup the requirements, but it failed. So I tried my luck ;-)

How can I set up the devcontainer to fix those issues?

**Comment by sven-ruess** [comment]
After some research I got all tools locally to run. I could fix the last issues and now I hope that this enhancement is good enough to get integrated in a further version.

**Comment by chemelli74** [comment]
> Hi @chemelli74, what built-in devcontainer? I tried to setup the requirements, but it failed. So I tried my luck ;-)
> 
> How can I set up the devcontainer to fix those issues?

[VScode -> Clone repo in ](https://code.visualstudio.com/docs/devcontainers/containers)

**Comment by chemelli74** [comment]
> After some research I got all tools locally to run. I could fix the last issues and now I hope that this enhancement is good enough to get integrated in a further version.

Thx for your effort; but unfortunately you are not using pre-commit config from this repo.
You miss commitlint checks.

**Comment by sven-ruess** [comment]
Hi @chemelli74, is this last issue what I need to fix? I think it is looking in wrong repo for this branch. Branch is available in sven-ruess/aiovodafone not chemelli74/aiovodafone, if I understand the output correct.

**Comment by sven-ruess** [comment]
Hi. Can you give me a short update? I think the last issue is not related to a mistake from me, or?

**Comment by chemelli74** [comment]
> Hi. Can you give me a short update? I think the last issue is not related to a mistake from me, or?

Not sure where the problem lies, need to investigate.

File: library_test.py [hunk 1]
```diff
@@ -92,19 +92,19 @@ async def main() -> None:
         try:
             await api.login()
         except ModelNotSupported:
-            print("Model is not supported yet for router", api.host)
+            print(f"Model is not supported yet for router {api.host}")
```
Comment (chemelli74) [inline library_test.py:95]: Please avoid un-needed changes, limit this PR to functional code.

File: library_test.py [hunk 2]
```diff
@@ -114,15 +114,46 @@ async def main() -> None:
 
     print("-" * 20)
     devices = await api.get_devices_data()
-    print("Devices:", devices)
+    print(f"Devices: {devices}")
+
     print("-" * 20)
     data = await api.get_sensor_data()
-    print("Data:", data)
```
Comment (chemelli74) [inline library_test.py:120]: The aim of the `library_test` script is to check that data retrieval works, not to be used as stand-alone application.
So for me it doesn't make sense to format output.

File: src/aiovodafone/api.py [hunk 3]
```diff
@@ -168,15 +168,15 @@ async def _get_page_result(self, page: str) -> aiohttp.ClientResponse:
             allow_redirects=False,
         )
 
-    @abstractmethod
```
Comment (chemelli74) [inline src/aiovodafone/api.py:171]: Please avoid un-needed changes, limit this PR to functional code.

File: src/aiovodafone/api.py [hunk 4]
```diff
@@ -230,9 +238,15 @@ async def _encrypt_string(
...
+    def convert_uptime(self, uptime: str) -> datetime:
+        """Convert uptime to datetime."""
+        return datetime.now(tz=UTC) - timedelta(
+            seconds=int(uptime),
+        )
+
     async def login(self) -> bool:
         """Router login."""
-        _LOGGER.debug("Logging into %s", self.host)
+        _LOGGER.debug(f"Logging into {self.host}")
```
Comment (chemelli74) [inline src/aiovodafone/api.py:249]: Please avoid un-needed changes, limit this PR to functional code.

File: src/aiovodafone/api.py [hunk 5]
```diff
@@ -278,11 +292,12 @@ async def get_devices_data(self) -> dict[str, VodafoneStationDevice]:
 
         """
         _LOGGER.debug("Get hosts")
-        host_response = await self._get_page_result("/api/v1/host/hostTbl")
-        host_json = await host_response.json()
```
Comment (chemelli74) [inline src/aiovodafone/api.py:282]: Please avoid un-needed changes, limit this PR to functional code.

File: src/aiovodafone/api.py [hunk 6]
```diff
@@ -308,22 +323,92 @@ async def get_devices_data(self) -> dict[str, VodafoneStationDevice]:
 
     async def get_sensor_data(self) -> dict[Any, Any]:
         """Get all sensors data."""
-        status_response = await self._get_page_result("/api/v1/sta_status")
-        status_json = await status_response.json()
```
Comment (chemelli74) [inline src/aiovodafone/api.py:312]: Please avoid un-needed changes, limit this PR to functional code.

File: src/aiovodafone/api.py [hunk 7]
```diff
@@ -308,22 +323,92 @@ async def get_devices_data(self) -> dict[str, VodafoneStationDevice]:
...
+        data["wan_status"] = response_json["data"]["WANStatus"]
+        data["cm_status"] = response_json["data"]["CMStatus"]
+        data["cm_mode"] = response_json["data"]["LanMode"]
+        data["sys_serial_number"] = response_json["data"]["serialnumber"]
+        data["sys_firmware_version"] = response_json["data"]["firmwareversion"]
+        data["sys_hardware_version"] = response_json["data"]["hardwaretype"]
+        data["sys_uptime"] = response_json["data"]["uptime"]
         return data
 
-    def convert_uptime(self, uptime: str) -> datetime:
```
Comment (chemelli74) [inline src/aiovodafone/api.py:322]: Please avoid un-needed changes, limit this PR to functional code.

**Review by chemelli74 — DISMISSED** [review]
Please review the whole PR and limit changes to functional code.

The rest should go, if really wanted, in a different PR.

**Comment by chemelli74** [comment]
> All points were reworked.

Thx, everything seems ok !

Will test the new code with my router and if it works, I'll merge.

**Comment by sven-ruess** [comment]
That is great. I am using this for the last two month for monitoring my cable modem. It is perfectly working. Also I integrated it in my local Home Assistant installation. There it is also working.

Thank you for your support and I think, I will prepare some further pull requests for rework and Python3 features.

File: src/aiovodafone/api.py [hunk 8]
```diff
@@ -317,6 +325,9 @@ async def get_sensor_data(self) -> dict[Any, Any]:
         data["sys_firmware_version"] = status_json["data"]["firmwareversion"]
         data["sys_hardware_version"] = status_json["data"]["hardwaretype"]
         data["sys_uptime"] = status_json["data"]["uptime"]
+        data["wan_status"] = status_json["data"]["WANStatus"]
+        data["cm_status"] = status_json["data"]["CMStatus"]
+        data["cm_mode"] = status_json["data"]["LanMode"]
```
Comment (chemelli74) [inline src/aiovodafone/api.py:330]: Another pass or review, showed that line. Why ?

```suggestion
        data["lan_mode"] = status_json["data"]["LanMode"]
```

File: library_test.py [hunk 9]
```diff
@@ -117,12 +117,21 @@ async def main() -> None:
...
+    print("Sensor Data:", data)
     print("-" * 20)
     print("Serial #:", data["sys_serial_number"])
     print("Firmware:", data["sys_firmware_version"])
     print("Hardware:", data["sys_hardware_version"])
     print("Uptime  :", api.convert_uptime(data["sys_uptime"]))
+    print("WAN status:", data.get("wan_status"))
+    print("Cable modem status:", data.get("cm_status"))
+    print("Cable modem mode:", data.get("cm_mode"))
```
Comment (chemelli74) [inline library_test.py:128]:
```suggestion
    print("LAN mode:", data.get("lan_mode"))
```

File: src/aiovodafone/api.py (hunk 8 above)
Comment (sven-ruess) [reply src/aiovodafone/api.py:330]: LanMode describes the functionality of cable modem. There are two mods available:

- router
- bridge-static

Those mods describe the functionality, which will be done by modem. Router is acting as a classical router. Bridge-static is acting a bridge device. In GUI it is named as modem mode. The decision from my side was, to take this naming. LAN mode can be misunderstood.

You can decide how the naming should be. It is your project. I am happy, with your decision.

File: library_test.py (hunk 9 above)
Comment (sven-ruess) [reply library_test.py:128]: It is related to above change. Further change to get it working.

Feel free to choice your wording.

File: src/aiovodafone/api.py [hunk 10]
```diff
@@ -325,6 +336,78 @@ def convert_uptime(self, uptime: str) -> datetime:
...
+
+        data = {"downstream": {}, "upstream": {}}  # type: dict[Any, Any]
+
+        # OFDM Downtream
+        for channel in response_json["data"]["ofdm_downstream"]:
+            data["downstream"][channel["channelid_ofdm"]] = {
+                "channel_type": channel["ChannelType"],
+                "channel_frequency": channel["start_frequency"],
+                "channel_modulation": channel["FFT_ofdm"],
+                "channel_signal": channel["power_ofdm"],
```
Comment (chemelli74) [inline src/aiovodafone/api.py:354]:
```suggestion
                "channel_power": channel["power_ofdm"],
```

File: src/aiovodafone/api.py [hunk 11]
```diff
@@ -325,6 +336,78 @@ def convert_uptime(self, uptime: str) -> datetime:
...
+                "channel_locked": channel["locked_ofdm"],
+            }
+
+        # Downtream
+        for channel in response_json["data"]["downstream"]:
+            data["downstream"][channel["channelid"]] = {
+                "channel_type": channel["ChannelType"],
+                "channel_frequency": channel["CentralFrequency"],
+                "channel_modulation": channel["FFT"],
+                "channel_signal": channel["power"],
```
Comment (chemelli74) [inline src/aiovodafone/api.py:364]:
```suggestion
                "channel_power": channel["power"],
```

File: src/aiovodafone/api.py [hunk 12]
```diff
@@ -325,6 +336,78 @@ def convert_uptime(self, uptime: str) -> datetime:
...
+                "channel_locked": channel["locked"],
+            }
+
+        # OFDMA upstream
+        for channel in response_json["data"]["ofdma_upstream"]:
+            data["upstream"][channel["channelidup"]] = {
+                "channel_type": channel["ChannelType"],
+                "channel_frequency": channel["start_frequency"],
+                "channel_modulation": channel["FFT"],
+                "channel_signal": channel["power"],
```
Comment (chemelli74) [inline src/aiovodafone/api.py:374]:
```suggestion
                "channel_power": channel["power"],
```

File: src/aiovodafone/api.py [hunk 13]
```diff
@@ -325,6 +336,78 @@ def convert_uptime(self, uptime: str) -> datetime:
...
+                "channel_locked": channel["RangingStatus"],
+            }
+
+        # Upstream
+        for channel in response_json["data"]["upstream"]:
+            data["upstream"][channel["channelidup"]] = {
+                "channel_type": channel["ChannelType"],
+                "channel_frequency": channel["CentralFrequency"],
+                "channel_modulation": channel["FFT"],
+                "channel_signal": channel["power"],
```
Comment (chemelli74) [inline src/aiovodafone/api.py:384]:
```suggestion
                "channel_power": channel["power"],
```

File: src/aiovodafone/api.py [hunk 14]
```diff
@@ -325,6 +336,78 @@ def convert_uptime(self, uptime: str) -> datetime:
...
+        _LOGGER.debug("Get voice data")
+        response = await self._get_page_result("/api/v1/sta_voice_status")
+        response_json = await response.json()
+        _LOGGER.debug(f"GET reply ({response_json})")
+
+        data = {"line1": {}, "line2": {}}  # type: dict[Any, Any]
+
+        for line in ["1", "2"]:
+            data[f"line{line}"] = {
+                "line_number": response_json["data"][f"callnumber{line}"],
```
Comment (chemelli74) [inline src/aiovodafone/api.py:402]:
```suggestion
                "call_number": response_json["data"][f"callnumber{line}"],
```

**Comment by chemelli74** [comment]
@sven-ruess please confirm latest code (just updated) works fine for you so I can merge it.

Thx!

**Review by chemelli74 — APPROVED** [review]

**Comment by sven-ruess** [comment]
Latest code update is working for me. I tested bridge and router mode.

**Pull request merged**